File: rateLogger.js
const map = {};
const logger = (timeStamp, message) => {
  if (map[message] === undefined) {
    map[message] = timeStamp;
    return true;
  } else {
    if (timeStamp - map[message] < 10) {
      return false;
    }
    map[message] = timeStamp;
    return true;
  }
};

console.log(logger(1, "foo"));
console.log(logger(2, "bar"));
console.log(logger(3, "foo"));
console.log(logger(8, "bar"));
console.log(logger(10, "foo"));
console.log(logger(11, "foo"));

File: Rate_Logger.java
// TC - O(1) 
// SC - O(n)  
class Logger {
    HashMap<String, Integer> map;
    public Logger() {
        map = new HashMap<>();
    }


    public boolean shouldPrintMessage(int timestamp, String message) {
        if(!map.containsKey(message)) {
              map.put(message, timestamp); return true;  
        }

        if(map.get(message) + 10 <= timestamp) {
            map.put(message, timestamp);  
            return true;
        }
        return false; 
    }
}
